src/arkode/arkode_mristep.c

```diff
@@ -1082,42 +1078,65 @@ int mriStep_Init(void* arkode_mem, int init_type)
                       "Error in coupling table");
       return (ARK_ILL_INPUT);
     }
-
+    
     /* stage types */
-    if (step_mem->stagetypes)
+    step_mem->stagetypes = realloc(step_mem->stagetypes,
+                                   step_mem->stages * sizeof(*step_mem->stagetypes));
+    if (step_mem->stagetypes == NULL)
     {
-      free(step_mem->stagetypes);
-      step_mem->stagetypes = NULL;
-      ark_mem->liw -= step_mem->stages;
+      arkProcessError(ark_mem, ARK_MEM_FAIL, __LINE__, __func__, __FILE__,
+                      MSG_ARK_MEM_FAIL);
+      return (ARK_MEM_FAIL);
     }
-    step_mem->stagetypes = (int*)calloc(step_mem->stages, sizeof(int));
-    ark_mem->liw += step_mem->stages;
+    ark_mem->liw += delta_stages;
     for (j = 0; j < step_mem->stages; j++)
     {
       step_mem->stagetypes[j] = mriStepCoupling_GetStageType(step_mem->MRIC, j);
     }
-
+    
     /* explicit RK coefficient row */
-    if (step_mem->Ae_row)
+    step_mem->Ae_row = realloc(step_mem->Ae_row,
+                               step_mem->stages * sizeof(*step_mem->Ae_row));
+    if (step_mem->Ae_row == NULL)
     {
-      free(step_mem->Ae_row);
-      step_mem->Ae_row = NULL;
-      ark_mem->lrw -= step_mem->stages;
+      arkProcessError(ark_mem, ARK_MEM_FAIL, __LINE__, __func__, __FILE__,
+                      MSG_ARK_MEM_FAIL);
+      return (ARK_MEM_FAIL);
     }
-    step_mem->Ae_row = (sunrealtype*)calloc(step_mem->stages,
-                                            sizeof(sunrealtype));
-    ark_mem->lrw += step_mem->stages;
+    ark_mem->lrw += delta_stages;
 
     /* implicit RK coefficient row */
-    if (step_mem->Ai_row)
+    step_mem->Ai_row = realloc(step_mem->Ai_row,
+                               step_mem->stages * sizeof(*step_mem->Ai_row));
+    if (step_mem->Ai_row == NULL)
     {
-      free(step_mem->Ai_row);
-      step_mem->Ai_row = NULL;
-      ark_mem->lrw -= step_mem->stages;
+      arkProcessError(ark_mem, ARK_MEM_FAIL, __LINE__, __func__, __FILE__,
+                      MSG_ARK_MEM_FAIL);
+      return (ARK_MEM_FAIL);
+    }
+    ark_mem->lrw += delta_stages;
+
+    /* Allocate reusable arrays for fused vector interface */
+    step_mem->nfusedopvecs = 2 * step_mem->stages + 2;
+    step_mem->cvals = realloc(step_mem->cvals,
+                              step_mem->nfusedopvecs * sizeof(*step_mem->cvals));
+    if (step_mem->cvals == NULL)
+    {
+      arkProcessError(ark_mem, ARK_MEM_FAIL, __LINE__, __func__, __FILE__,
+                      MSG_ARK_MEM_FAIL);
+      return (ARK_MEM_FAIL);
+    }
+    ark_mem->lrw += 2 * delta_stages;
+
+    step_mem->Xvecs = realloc(step_mem->Xvecs,
+                              step_mem->nfusedopvecs * sizeof(*step_mem->Xvecs));
```

Comment:
I'm okay with changing to `realloc`, but (a) please cast the result to the correct type (`N_Vector*`), and (b) `memset` the new memory to all `NULL` to retain `calloc`'s initialization feature.